sdk/identity/Azure.Identity/src/ManagedIdentityClient.cs

```diff
@@ -0,0 +1,419 @@
+
+            request.UriBuilder.AppendQuery("resource", Uri.EscapeDataString(resource));
+
+            if (!string.IsNullOrEmpty(clientId))
+            {
+                request.UriBuilder.AppendQuery("client_id", Uri.EscapeDataString(clientId));
+            }
+
+            return request;
+        }
+
+        private Request CreateCloudShellAuthRequest(string[] scopes, string clientId)
+        {
+            // covert the scopes to a resource string
+            string resource = ScopeUtilities.ScopesToResource(scopes);
+
+            Request request = _pipeline.CreateRequest();
+
+            request.Method = HttpPipelineMethod.Post;
+
+            request.Headers.Add(HttpHeader.Common.FormUrlEncodedContentType);
+
+            request.UriBuilder.Uri = s_endpoint;
+
+            request.Headers.Add("Metadata", "true");
+
+            var bodyStr = $"resource={Uri.EscapeDataString(resource)}";
+
+            if (!string.IsNullOrEmpty(clientId))
+            {
+                bodyStr += $"&client_id={Uri.EscapeDataString(clientId)}";
+            }
+
+            ReadOnlyMemory<byte> content = Encoding.UTF8.GetBytes(bodyStr).AsMemory();
+
+            request.Content = HttpPipelineRequestContent.Create(content);
+
+            return request;
+        }
+
+        private async Task<AccessToken> DeserializeAsync(Stream content, CancellationToken cancellationToken)
+        {
+            using (JsonDocument json = await JsonDocument.ParseAsync(content, default, cancellationToken).ConfigureAwait(false))
+            {
+                return Deserialize(json.RootElement);
+            }
+        }
+
+        private AccessToken Deserialize(Stream content)
+        {
+            using (JsonDocument json = JsonDocument.Parse(content))
+            {
+                return Deserialize(json.RootElement);
+            }
+        }
+
+        private AccessToken Deserialize(JsonElement json)
+        {
+            string accessToken = null;
+
+            DateTimeOffset expiresOn = DateTimeOffset.MaxValue;
+
+            if (json.TryGetProperty("access_token", out JsonElement accessTokenProp))
+            {
+                accessToken = accessTokenProp.GetString();
+            }
+
+            if (json.TryGetProperty("expires_on", out JsonElement expiresOnProp))
+            {
+                // if s_msiType is AppService expires_on will be a string formatted datetimeoffset
+                if (s_msiType == MsiType.AppService)
+                {
+                    expiresOn = DateTimeOffset.Parse(expiresOnProp.GetString());
+                }
+                // otherwise expires_on will be a unix timestamp seconds from epoch
+                else
+                {
+                    expiresOn = DateTime.UtcNow + TimeSpan.FromSeconds(expiresOnProp.GetInt64());
+                }
+            }
+
+            return new AccessToken(accessToken, expiresOn);
```

Comment:
What happens id access_token does not exist in the JSON payload? i.e. is it ok to pass null here?

Reply:
If the access_token to not exist here then the Token property on AccessToken will be null and this will be handled higher up in the stack.  So while it didn't get a token, it should not throw an exception from here.